net/sched/sch_gkprio.c

```diff
@@ -0,0 +1,298 @@
+/*
+ * net/sched/sch_gkprio.c  Gatekeeper Priority Queue.
+ *
+ *		This program is free software; you can redistribute it and/or
+ *		modify it under the terms of the GNU General Public License
+ *		as published by the Free Software Foundation; either version
+ *		2 of the License, or (at your option) any later version.
+ *
+ * Authors:	Nishanth Devarajan, <ndev_2021@gmail.com>
+ *	        original idea by Michel Machado, Cody Doucette, and Qiaobin Fu
+ */
+
+#include <linux/string.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/types.h>
+#include <linux/kernel.h>
+#include <linux/errno.h>
+#include <linux/skbuff.h>
+#include <net/pkt_sched.h>
+#include <net/sch_generic.h>
+#include <net/inet_ecn.h>
+
+/* Packets are assigned priorities [0, 63] due to the IP DSCP field limits. */
+#define GKPRIO_MAX_PRIORITY 64
+
+/*	  Gatekeeper Priority Queue
+ *	=================================
+ *
+ * This qdisc schedules a packet according to the value (0-63) of its DSCP
+ * (IPv4) or DS (IPv6) field, where a higher value places the packet closer
+ * to the exit of the queue. Non-IP packets are assigned a default priority
+ * specified to Gatekeeper; if none is specified, default priority is set 
+ * to 0. When the queue is full, the lowest priority packet in the queue is 
+ * dropped to make room for the packet to be added if it has higher priority.
+ * If the packet to be added has lower priority than all packets in the queue,
+ * it is dropped.
+ *
+ * Without the Gatekeeper priority queue, queue length limits must be imposed
+ * for individual queues, and there is no easy way to enforce a global queue
+ * length limit across all priorities. With the Gatekeeper queue, a global
+ * queue length limit can be enforced while not restricting the queue lengths
+ * of individual priorities.
+ *
+ * This is especially useful for a denial-of-service defense service,
+ * Gatekeeper, which prioritizes packets in flows that demonstrate expected
+ * behavior of legitimate users. The queue is flexible to allow any number
+ * of packets of any priority up to the global limit of the scheduler
+ * without risking resource overconsumption by a flood of low priority packets.
+ *
+ * The Gatekeper standalone codebase is found here:
+ *
+ *         	https://github.com/AltraMayor/gatekeeper
+ */
+
+struct gkprio_sched_data {
+	/* Parameters. */
+	u32 max_limit;
+	u16 noip_dfltp;
+
+	/* Queue state. */
+	struct qdisc_skb_head qdiscs[GKPRIO_MAX_PRIORITY];
```

Comment:
Hi Michel, I'm really not sure whether we should replace the data structure here. No other net/sched scheduler seems to be using `sk_buff_head` while many schedulers use `qdisc_skb_head` and it even being defined in `sch_generic.h`, I think its strongly suggested that we use `qdisc_skb_head` instead of `sk_buff_head` (please look at the other comment I made before this). Let me know what you think. Thanks.

Reply:
Is there a scheduler that uses `struct qdisc_skb_head` **and** adds to the tail of the list? Since that the current kernel doesn't have a function to add to the tail of `struct qdisc_skb_head`, finding that example would clarify what we could do. I don't think there's such example, though. If you are still not convinced of the need to change, you can try to find a scheduler that adds to the tail, and see what `struct` it uses.

Reply:
I looked through the schedulers and yes, htb uses the `struct qdisc_skb_head`  :  'direct queue' to add to the tail using its static `htb_enqueue_tail` function.

Reply:
I made a mistake in my previous comment. What I meant by "and adds to the tail of the list" was really "and drops at the tail of the list" since adding to the tail doesn't require the previous node pointer, but dropping at the tail of a list, as we need in our scheduler, requires. `htb_enqueue_tail(), which you pointed out, doesn't update the previous pointer for example.

Reply:
Hi Michel, as I mentioned in another comment the `skb->prev` is still defined, it is inherently a part of the `sk_buff` construct and not the `qdic_skb_head` construct. Our code currently does drop at the tail: loc 151 which is the code to drop packets at the tail of the list in the enqueue function, updating the prev ptr in the process. 

This functionality was  tested in one of the code reviews with Cody, and it did work functionally well (expected good results were obtained in those tests). Let me know what you think. Thanks.

Reply:
@ndev2, my goal is to make your code as well aligned with the kernel as possible as to minimize the chance of rejection during the review phase. The more your code reconstructs code present elsewhere or digresses from what is expected from it, the more you're going to hear complains during the review.

On the other hand, I feel I can't make it as clear to you as it is to me. I don't want to you force you to do what you don't agree or understand, so I think you should pursue the merge upstream using your understanding how to do so. Good luck.

Reply:
Ok, I was not clear on what was meant by making `skb->prev`  'undefined'. I apologise for the misunderstanding, and I will make the change to `sk_buff_head` as I understand that you can't update `skb->prev` in `qdisc_skb_head` code. Thank you for your comments. Hope you can look past my mistake, and we can work together on this.

Reply:
@ndev2, I want to work with you, but you must commit yourself to teamwork. If teamwork is not your thing, you may want to give up on coding now while you still have time to find and pursue a solitary activity. Because not being a good team player is going to limit yourself to small pieces of software.

I'm not abruptly saying these things. Not only is this topic quite old -- we have dissussed it back in August 2017 https://github.com/ndev2/net-next/pull/2#discussion_r134526042 -- it is the second time that I point out a problem in your code that must be addressed and, instead of trying to understand my point and move on, you come up with unsound arguments and waste everyone's time. The first time was when I pointed out an invalid memory access in an early version of this scheduler, and you insisted that it was correct justifying that it didn't triggered an error during tests.

I'm not saying this to make you accept whatever I say blindly because this behavor would ultimately doom our project. Nevertheless, if you stick with us, you have to personally grow to make it worth it for everyone, including yourself.

Reply:
Yes, I understand what you mean, it is a an issue that I've discovered only recently, but it is apparent that its been a problem for a long time now. I'm working on it and trying to be better (outside of coding as well). I'm grateful for your patience, will work on the code to make it good for the kernel. Thanks.